### DISCUSSION TITLE
Why is the MediaElement not playing audio files stored on my Google Drive?

### DISCUSSION THREAD

**Discussion by Riffy135** [CommunityToolkit/Maui#1203] [Q&A]
@

I have Audio Files (MP3/MP4) files that are stored on my Google Drive not being played when using the MediaElement of the MAUI Community Toolkit.

I am implementing the MediaElement functionality as follows:

In my XAML I have the following:
`<toolkit:MediaElement x:Name="mediaElement" Grid.Row="0" Source="{Binding MediaFile}" ShouldAutoPlay="True" /\>`

The MediaElement is defined as follows in my xaml.cs code:
`

public partial class Level3Page : ContentPage
{
    readonly Level3ViewModel _viewModel;
     
    public static MediaElement MediaControlA { get; private set; }

    public Level3Page()
    {
        InitializeComponent();
        
        MediaControlA = mediaElement;
       
        BindingContext = _viewModel = new Level3ViewModel(this);
    } 
}`

I am using a Google Service Account for connecting my App to the Google Drive folder. This all works fine and also the File to play is located correctly.  But MediaElement does not play the audio and I get the following error:

0:] MediaManager: Error: Error codename: ERROR_CODE_PARSING_CONTAINER_UNSUPPORTED, Error message: Source error

The following is my C# code in my ViewModel to get the file details from Google Drive and using the MediaElement Play method: Level3Page.MediaControlA.Play():

`
         
           request = Services.Files.List();
            request.Q = $"mimeType!='application/vnd.google-apps.folder' and name='{fileName}' and '{folder.Id}' in parents";

            request.Fields = "files(id,name)";
            result = await request.ExecuteAsync();
            if (result.Files != null && result.Files.Any())
            {
                var FName1 = result.Files.First().Name;
                MediaFile = "https://drive.google.com/file/d/" + result.Files[0].Id;
                Level3Page.MediaControlA.Play();
            }`

I know that the value of result.Files[0].Id is correct, because when I copy and paste the MediaFile URL contents into my web browser (have to login to google drive), the audio file plays without any issue.

The url example is:
https://drive.google.com/file/d/MY FILE ID

Can someone please suggest what the issue could be?

Thanks

**Comment by VladislavAntonyuk** [comment]
You should provide a direct link to file. Assume you are in a private mode and pass the url. Google asks you to sign in. The same happens with media element. You should either specify a direct link to file or download it and play locally

**Reply by Riffy135** [reply]
Thanks for response Antony.

The  result.Files[0].Id set in MediaElement is a direct link which is returned after I sign in with my Service Account. As the App is already signed in, I don't think I need to sign in again through the App.

I was demonstrating that the File Id is correct for the Audio file, as it will play in Web browser outside of my App.

**Reply by VladislavAntonyuk** [reply]
Try your link in browser private mode. If it works then media element should play it. Could you share the link for a demo media file?

**Reply by Riffy135** [reply]
The sample file id is 1bpRH1OVmz8tke5l-reagePDQG-cesMRB.

The url is: https://drive.google.com/file/d/1bpRH1OVmz8tke5l-reagePDQG-cesMRB. Using this in private mode prompts me for Google drive user/password.

I know the file id is correct because, as a test,  I use it to download the file from Google drive on to local file within the App, I then populate MediaFile with local filename and using MediaElement to play the audio file. This works fine with several files I tested it.

Downloading the file is not ideal for me, as it creates a few seconds (noticable) delay for the user before the audio starts. Also I want to play some videos from the Google drive, which would be impractical to download first, due to the file sizes.

Ideally I want MediaElement to stream/play the files directly from Google Drive. So if you are able to identify reason why the url is not being accepted as a good Source, it would be great.

**Reply by pictos** [reply]
@Riffy135 can you try to format your url to match this scheme?
https://stackoverflow.com/a/69817790

**Reply by Riffy135** [reply]
I still get [0:] MediaManager: Error: Error codename: ERROR_CODE_PARSING_CONTAINER_UNSUPPORTED, Error message: Source error, when now setting MediaFile to 

https://drive.google.com/uc?id=1bpRH1OVmz8tke5l-reagePDQG-cesMRB

or

https://drive.google.com/uc?id=1bpRH1OVmz8tke5l-reagePDQG-cesMRB&export=download.

Eg Level3Page.MediaControlA.Source.Uri.AbsoluteUri.AbsoluteUri = https://drive.google.com/uc?id=1bpRH1OVmz8tke5l-reagePDQG-cesMRB&export=download.

**Reply by VladislavAntonyuk** [reply]
All these urls require authentication, that is why media element can't play them.

**Reply by Riffy135** [reply]
The app is already authenticating when getting the file id from Google Drive, so why can MediaElement not use that same connection to retrieve the url?

**Reply by pictos** [reply]
@Riffy135 you probably need to add some info on the request, like some token or header. I would say to look into android samples on how to do that, then you just need to translate the solution to C#

**Comment by CliffAgius** [comment]
Closed and answered
